src/main/java/com/amazon/ion/bytecode/BytecodeIonReader.kt

```diff
@@ -2,13 +2,612 @@
+                OperationKind.BOOL,
+                OperationKind.INT,
+                OperationKind.FLOAT,
+                OperationKind.DECIMAL,
+                OperationKind.TIMESTAMP,
+                OperationKind.STRING,
+                OperationKind.SYMBOL,
+                OperationKind.BLOB,
+                OperationKind.CLOB,
+                OperationKind.LIST,
+                OperationKind.SEXP,
+                OperationKind.STRUCT -> break
+                OperationKind.END -> {
+                    // TODO(perf): See if there is any benefit to splitting these into different operation kinds.
+                    when (instruction) {
+                        // return so that we don't update the current state so that if the user keeps calling `next()` they keep encountering this.
+                        Instructions.I_END_CONTAINER -> return null
+                        // break so that we do update the current state so that if the user calls `next()`, we can check the generator for more data.
+                        Instructions.I_END_OF_INPUT -> break
+                        else -> TODO("${Debugger.renderSingleInstruction(instruction)} at ${i - 1}")
+                    }
+                }
+                OperationKind.FIELD_NAME -> fieldNameIndex = i - 1
+                OperationKind.ANNOTATIONS -> {
+                    annotationStart = (annotationStart or (i and annotationFlag))
+                    annotationCount++
+                    annotationFlag = 0
+                }
+                OperationKind.IVM -> handleIvm(instruction)
+                OperationKind.REFILL -> {
+                    bytecode = refillBytecode()
+                    i = 0
+                }
+                OperationKind.DIRECTIVE -> {
+                    i = handleSystemValue(instruction, i)
+                }
+                OperationKind.ARGUMENT -> {
+                    // In other words, an absent argument.
+                    annotationStart = 0
+                    annotationFlag = -1
+                    annotationCount = 0
+                }
+                OperationKind.METADATA -> {
+                    labelIndex = i - 1
+                    continue
+                }
+                else -> TODO("${OperationKind.nameOf(operationKind)} at ${i - 1}")
+            }
+        } while (true)
+
+        if (labelIndex != -1) {
+            this.labelIndex = labelIndex
+        }
+
+        this.bytecodeI = i
+        this.instruction = instruction
+        this.annotationsIndex = annotationStart - 1
+        this.annotationCount = annotationCount.toByte()
+
+        return OperationKind.ionTypeOf(operationKind)
+    }
+
+    private fun handleIvm(instruction: Int) {
+        val ionVersionInt = Instructions.getData(instruction)
+        minorVersion = ionVersionInt.toByte()
+        generator = generator.getGeneratorForMinorVersion(ionVersionInt)
+        symbolTable = EncodingContextManager.SYSTEM_SYMBOLS
```

Comment:
We should be clearing the constant pool here too if we are managing it in the writer, yes? Can avoid unnecessary resizing.

```suggestion
        symbolTable = EncodingContextManager.SYSTEM_SYMBOLS
        constantPool.clear()
```